Fix part of #13162: Add schema to validate StoryUrlFragmentHandler class arguments and a Submit button for story url fragment

SanjaySajuJacob opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of [#13162](https://github.com/oppia/oppia/issues/13162).
2. This PR provides schema for StoryUrlFragmentHandler class and a submit button for story url fragment during story creation.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The linter/Karma presubmit checks have passed locally on your machine.
- [ ] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [ ] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

https://user-images.githubusercontent.com/70763993/144430978-b2059f4a-7414-4677-b026-bff49e8bb5d0.mp4

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

SanjaySajuJacob committed:
Schema for  StoryUrlFragmentHandler

SanjaySajuJacob committed:
Merge branch 'schema-validation' of https://github.com/SanjaySajuJacob/oppia into schema-validation

vojtechjelinek reviewed (approved):
LGTM!

SanjaySajuJacob commented:
@aks681 are there any changes that I need to make since 4 tests are failing?

aks681 reviewed (approved):
Lgtm as codeowner!

aks681 commented:
@SanjaySajuJacob Actually, it seems like that the second test failure is not a flake and might be due to your change:
![image](https://user-images.githubusercontent.com/22347970/143153657-bed7856e-9e99-432a-81ae-58f960e7c832.png)

Can you try the steps locally, that are present in the e2e test? So, basically create a topic, and add a story to it in the topic editor (fill out dummy details for story in the story editor as well). Somewhere along this flow is when the error occurs. Feel free to ping me or ask here if you have any questions on accessing these pages!

aks681 reviewed (changes requested):
Unapproving due to failing e2e tests

SanjaySajuJacob commented:
![error1](https://user-images.githubusercontent.com/70763993/143179137-cbe7e494-7532-4d52-9bb2-acdfbf135b11.PNG)

@aks681 I tried the steps you had instructed and I came across this error while giving the story url fragment. I think this error occurs because the regex pattern that validates the url fragment does not support those which end in '-' (hyphen). I think this is the same error as in the e2e tests. What changes would you suggest I make in this case? Should the regular expression be changed to allow url fragments which end in a hyphen also?

SanjaySajuJacob commented:
The full story url fragment I gave was "test-story", this error occurred as I filled in "test-" into story url fragment

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

SanjaySajuJacob committed:
Merge branch 'develop' of https://github.com/oppia/oppia into schema-validation

SanjaySajuJacob committed:
Merge branch 'schema-validation' of https://github.com/SanjaySajuJacob/oppia into schema-validation

aks681 commented:
Hi @SanjaySajuJacob. Ah ok, in that case, can you modify the test to not include hyphen? Since the regex is supposed to be the correct representation of the fragment.

SanjaySajuJacob commented:
Hi @aks681, I understand that the e2e tests work in a way that they simulate typing from a user. I am not certain this is how the tests work, would you please tell me if this is correct? 
I see that this [e2e test](https://github.com/oppia/oppia/runs/4332885736?check_suite_focus=true) gives errors at every hyphen ("topic-", "topic-and-", etc) where the full story url fragment is "topic-and-story-editor-one"

![e2e 1](https://user-images.githubusercontent.com/70763993/143689515-ecc982f0-b93b-4a38-9c26-eadcf3a9a734.PNG)

These error I am assuming will go if the hyphens are removed from the url fragment (make it as "topicandstoryeditorone").
Would you please tell me if this is a viable solution?

aks681 commented:
Yup, e2e tests verify the full frontend to backend control flow by simulating inputs from the user on the various pages and verifying the result.
Actually, that is what I thought initially, but the regex itself in `constants.SCHEMA_FOR_STORY_URL_FRAGMENTS` is `^[a-z]+(-[a-z]+)*$`, right? This regex does allow hyphens in the name, although you are right that it wouldn't support something with the hyphen at the end.
I think what's happening is, in the editor, everytime something is typed, a backend call is run to update that fragment, in which case, errors occur as you are typing the fragment with hyphens in it, but the end value is still valid. Can you verify this?

If so, the simplest way to fix this would be to add a 'submit' button in the editor beside this input field so that the backend call is done only after the user has typed in the full valid fragment. This would require some frontend changes in the topic and story editor files. Would you be able to do it? Since, we cannot merge this PR as it is, since it would lead to failed e2e tests (the tests themselves are correct).

SanjaySajuJacob commented:
@aks681 as I understand it, the checks (by validation schema) occur dynamically on the url fragment when searching if it already exists, as below (`core/templates/pages/topic-editor-page/modal-templates/create-new-story-modal.template.html`)

![1 t](https://user-images.githubusercontent.com/70763993/143765916-80a83bf6-f269-424b-a9e9-70a3f00a4764.PNG)

So I will have to add a button which performs the the above mentioned check for the existence of the url fragment, after it is clicked. Would this be the right way to implement the solution you had given? I am not clear regarding the syntax of this.

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

SanjaySajuJacob committed:
Adding submit button for story url fragment

SanjaySajuJacob committed:
Merge branch 'schema-validation' of https://github.com/SanjaySajuJacob/oppia into schema-validation

SanjaySajuJacob committed:
Resolving lint checks

SanjaySajuJacob commented:
Would someone please help me out with this? Is this error due to any issues in my code?

SanjaySajuJacob committed:
Fixing errors

SanjaySajuJacob commented:
Would someone help me out with this? Would you please tell me if the 'Invalid story id' error (in the Lighthouse CI tests) is something that I should alter or not?

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

aks681 commented:
Hi @SanjaySajuJacob. Sorry for the delay, can you assign the PR to me on github (and other reviewers) when you are waiting for response, that way we can get notified about it quicker.
I'm not sure what check you are referring to here. If it is the function called on `ng-change`, you're probably right. Check `onStoryUrlFragmentChange`. If that does a backend call, then that is the function that needs to be moved into a separate button so that backend calls aren't done on intermediate invalid fragments. Does that help?

SanjaySajuJacob commented:
Hi @aks681, I have removed `onStoryUrlFragmentChange` and added `assignUrlFragment` which is called on clicking the button and performs the checks on the entered story url fragment. I have added the video of its working under proof of this PR.

SanjaySajuJacob commented:
@aks681 PTAL

SanjaySajuJacob commented:
@aks681 PTAL

SanjaySajuJacob commented:
@kevintab95 @gp201 @U8NWXD PTAL

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

SanjaySajuJacob commented:
@aks681 @kevintab95 @gp201 Would you please take a look and inform me regarding any changes.

aks681 reviewed (approved):
Lgtm as codeowner! Though, e2e tests are still failing?

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

gp201 reviewed (approved):
LGTM

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

vojtechjelinek commented:
@SanjaySajuJacob The backend coverage tests are failing, you will need to write tests to cover the lines that are not tested now.

SanjaySajuJacob committed:
Merge branch 'develop' of https://github.com/oppia/oppia into schema-validation

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

SanjaySajuJacob committed:
Adding backend tests

SanjaySajuJacob committed:
Merge branch 'schema-validation' of https://github.com/SanjaySajuJacob/oppia into schema-validation

SanjaySajuJacob commented:
@vojtechjelinek PTAL would you please tell me if these Lighthouse CI test failures are something I have to alter? or can the PR be merged?

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

vojtechjelinek commented on core/templates/pages/topic-editor-page/modal-templates/create-new-story-modal.controller.spec.ts:
```diff
@@ -84,16 +84,16 @@ describe('Create New Story Modal Controller', function() {
     expect($scope.storyUrlFragmentExists).toBeFalse();
-    $scope.story.urlFragment = 'test-url';
-    $scope.onStoryUrlFragmentChange();
+    $scope.unsubmittedUrlFragment = 'test-url';
+    $scope.assignUrlFragment();
```
Why this change?

vojtechjelinek replied:
And ditto the stuff below.

SanjaySajuJacob replied:
@vojtechjelinek, I had added a button for submitting the url fragment and as a part of this I had removed the onStoryUrlFragmentChange() function and moved its functionality to assignUrlFragment() function which is called when the button is clicked. The unsubmittedUrlFragment is a temporary variable that holds the url fragment before it is submitted. I changed the frontend tests here accordingly.

Would this be fine?

vojtechjelinek replied:
Umm I don't think this is the right solution to the problem. Also if we were to do it this way we would need to do it in all places where we have a similar code.

SanjaySajuJacob replied:
@vojtechjelinek, I didnt really understand, is it that the frontend tests are not correct? (as in, it wont work) or is it that the coding style followed by me should be changed? Would you please give me some clarity into the issue.

SanjaySajuJacob replied:
> Umm I don't think this is the right solution to the problem. Also if we were to do it this way we would need to do it in all places where we have a similar code.

@vojtechjelinek,
 Do you mean adding the button is not the right solution to the problem? It was added because the end to end tests would otherwise fail. The hyphen in the url fragment gives error as the story editor fills it in. Therefore @aks681 had suggested that a button be added for submitting the final url fragment. 
What would you suggest to be an alternate solution?

vojtechjelinek replied:
I think the schema changes should not break the e2e tests, you should find as solution that works without introducing the button. Either modify the e2e tests or modifying the schema.

SanjaySajuJacob replied:
@vojtechjelinek, would there be other files that must be changed? Since story creation is present only in `core/templates/pages/topic-editor-page/modal-templates/create-new-story-modal.template.html` right.

SanjaySajuJacob replied:
@vojtechjelinek, a possible solution would be to remove the hyphens from the e2e tests, for example in the file `core/tests/protractor_desktop/learnerDashboard.js`, theres a test with url fragment "story-one" which if changed to "storyone" would allow the e2e tests to pass. 

Otherwise, the regex in the schema that checks the url fragment can be modified to allow url fragments that end with a hyphen to pass. 

Would either of these be viable solutions?

SanjaySajuJacob replied:
Also the thing is that, I think story creation is present in `core/templates/pages/topic-editor-page/modal-templates/create-new-story-modal.template.html` alone and not in any other file (the `onStoryUrlFragmentChange()` function also exists only in the `core/templates/pages/topic-editor-page/modal-templates/create-new-story-modal.controller.ts` file alone) as well as all the frontend, e2e and backend tests passed. 
So would there be any other file where changes have to be made? So we can consider the submit button as a fair solution right.

vojtechjelinek replied:
> @vojtechjelinek, a possible solution would be to remove the hyphens from the e2e tests, for example in the file `core/tests/protractor_desktop/learnerDashboard.js`, theres a test with url fragment "story-one" which if changed to "storyone" would allow the e2e tests to pass.
> 
> Otherwise, the regex in the schema that checks the url fragment can be modified to allow url fragments that end with a hyphen to pass.
> 
> Would either of these be viable solutions?

Yes, removing the hyphens seems fine.



> Also the thing is that, I think story creation is present in `core/templates/pages/topic-editor-page/modal-templates/create-new-story-modal.template.html` alone and not in any other file (the `onStoryUrlFragmentChange()` function also exists only in the `core/templates/pages/topic-editor-page/modal-templates/create-new-story-modal.controller.ts` file alone) as well as all the frontend, e2e and backend tests passed. So would there be any other file where changes have to be made? So we can consider the submit button as a fair solution right.

The problem is that topic or skill creation are quite similar to story creation and I think they should be pretty similar, but now we only use the button in the story creator.

SanjaySajuJacob replied:
>The problem is that topic or skill creation are quite similar to story creation and I think they should be pretty similar, but now we only use the button in the story creator.

Oh I see
>Yes, removing the hyphens seems fine.

Thanks, I will implement that as soon as possible

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

SanjaySajuJacob committed:
Merge branch 'oppia:develop' into schema-validation

tlidev committed:
Fix #13680 (#14463)

bhavukJain committed:
Fix part of #14219: Add a test to complete gae_models coverage (#14466)

chiragbaid7 committed:
Fix part of #14033: Type annotate improvements and feedback domain files (#14445)

Co-authored-by: Chirag Baid <chiragbaid@pop-os.localdomain>

heyimShivam committed:
Fix part of issue #9749: migrate preview-summary-tile-modal to angular component and some services (#14388)

* done

* done

* done html

* test case

* last commit from my end

* ashu changes

* extra

* ets

* remove line

* e2e test

* try to solve e2e

* eric

* resolved e2e

* resolved

* reset

* needed e2e changes

* needed e2e changes

* front end test

* front end test

SanjaySajuJacob committed:
Adding schema for StoryUrlFragmentHandler

SanjaySajuJacob committed:
Merge branch 'schema-validation' of https://github.com/SanjaySajuJacob/oppia into schema-validation

SanjaySajuJacob committed:
Removing previous changes

SanjaySajuJacob commented:
@vojtechjelinek, I dont understand what these changes are, I have only changed the story urls on few of the e2e tests and manually removed some of the changes that I had made before, since I could not figure out how to revert them and push. Are these changes from some other PRs? How do I revert back to the previous commit on my github branch?
(The only new changes I have made is the backend test for StoryUrlFragmentHandler, removal of hyphens from the url fragment for story creation in e2e tests and the schema for StoryUrlFragmentHandler)

SanjaySajuJacob commented:
Would someone please tell me why this has occurred and how I can revert commits on this github branch and add new changes to it.
Or should I just close this PR and make a new one?

vojtechjelinek commented:
> Would someone please tell me why this has occurred and how I can revert commits on this github branch and add new changes to it. Or should I just close this PR and make a new one?

Probably some git mistake, I recommend creating a new PR.
